File: tcp/send_all_then_close/cxx/tcp_streamer_server.cpp
#include <string>
#include <sstream>
#include <iostream>
#include <gflags/gflags.h>
#include "wrapsock.hpp"

void server(const char *host, uint16_t port) {
    Socket sock(AF_INET, SOCK_STREAM);
    sock.Setsockopt(SOL_SOCKET, SO_REUSEADDR, 1);
    sock.Bind(SocketAddress(host, port));
    sock.Listen(1);
    std::cout << "Listening at " << sock.Getsockname().ToString() << std::endl;
#if __cplusplus >= 201703L
    auto [sc, sockname] = sock.Accept(); // only for c++17
#else
    Socket sc;
    SocketAddress sockname;
    std::tie(sc, sockname) = sock.Accept();
#endif
    std::cout << "Accepted connection from " << sockname << std::endl;
    sc.Shutdown(SHUT_WR);
    std::string message;
    char buf[8192];
    while (true) {
        auto n = sc.Recv(buf, sizeof(buf));
        if (n == 0) {
            std::cout << "Received zero bytes - end of file" << std::endl;
            break;
        }
        std::cout << "Received " << n << " bytes" << std::endl;
        message.append(buf, n);
    }
    std::cout << "Message:\n" << std::endl;
    std::cout << message << std::endl;
    sc.Close();
    sock.Close();
}

std::string usage(const char* prog) {
    std::ostringstream os;
    os << "Send and receive over TCP\n\n"
        << "Usage: " << prog << " [--help] [--host HOST] [--port PORT]";
    return os.str();
}

DEFINE_string(host, "0.0.0.0", "IP address the server listens at");
DEFINE_int32(port, 1060, "TCP port number");

int main(int argc, char *argv[]) {
    gflags::SetUsageMessage(usage(argv[0]));
    gflags::ParseCommandLineFlags(&argc, &argv, true);

    server(FLAGS_host.c_str(), FLAGS_port);

    return 0;
}

File: tcp/send_all_then_close/python/tcp_streamer_server.py
#!/usr/bin/env python3

import socket
from argparse import ArgumentParser

def server(host, port):
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((host, port))
    sock.listen(1)
    print('Listening at', sock.getsockname())
    sc, sockname = sock.accept()
    print('Accepted connection from', sockname)
    sc.shutdown(socket.SHUT_WR)
    message = b''
    while True:
        more = sc.recv(8192)  # arbitrary value of 8k
        if not more:  # socket has closed when recv() returns ''
            print('Received zero bytes - end of file')
            break
        print('Received {} bytes'.format(len(more)))
        message += more
    print('Message:\n')
    print(message.decode('ascii'))
    sc.close()
    sock.close()

if __name__ == '__main__':
    parser = ArgumentParser(description='Transmit & receive a data stream')
    parser.add_argument('--host', type=str, default='0.0.0.0', help='IP address the server listens at (default: %(default)s)')
    parser.add_argument('--port', '-p', type=int, default=1060, help='TCP port number (default: %(default)s)')
    args = parser.parse_args()

    print("args: {}".format(args))

    server(args.host, args.port)
